File: JPP/lab-3/z2/src/Galois.java
public abstract class Galois {
    abstract public long order();
    protected long value;

    final public long get_value()
    {
        return value;
    }

    final public void set(long value)
    {
        this.value = value % order();
    }

    final public boolean equals(Galois other)
    {
        return this.compare(other) == 0;
    }

    final public long compare(Galois other)
    {
        return Long.compare(this.value, other.value);
    }

    final public void set(Galois other)
    {
        this.value = other.value;
    }

    static public Galois zero()
    {
        return null;
    };

    static public Galois one()
    {
        return null;
    };

    abstract public Galois add(Galois other);
    abstract public Galois sub(Galois other);
    abstract public Galois mul(Galois other);
    abstract public Galois div(Galois other);

    abstract public Galois inverse_add();
    abstract public Galois inverse_mul();

    final public String toString() {
        return Long.toString(value);
    }

    final protected long egcd(long a)
    {
        long b = order();

        long x = 1, y = 0;
        long x1 = 0, y1 = 1;

        while (b != 0) {
            final long q = a / b;

            final long x2 = x - q * x1;
            x = x1;
            x1 = x2;

            final long y2 = y - q * y1;
            y = y1;
            y1 = y2;

            final long temp = a - q * b;
            a = b;
            b = temp;
        }

        if (x < 0)
            x += order();

        return x;
    };
}

File: JFTT/lab-3/z2/utils/galois.py

class Galois:
    def __init__(this, P: int, value: int):
        this.P = P
        this.value = this._normalize(value)

    def _normalize(this, a: int) -> int:
        a %= this.P

        if a < 0:
            a += this.P
        
        return a
    
    def _inverse(this, a: int) -> int:
        b: int = this.P

        x: int = 1
        y: int = 0

        x1: int = 0
        y1: int = 1

        while b != 0:
            q: int = a // b
            
            x, x1 = x1, x - q * x1
            y, y1 = y1, y - q * y1
            a, b  = b,  a - q * b

        return this._normalize(x) 

    def add(this, other: 'Galois') -> 'Galois':
        return Galois(this.P, this.value + other.value)
    
    def sub(this, other: 'Galois') -> 'Galois':
        return Galois(this.P, this.value - other.value)
    
    def mul(this, other: 'Galois') -> 'Galois':
        return Galois(this.P, this.value * other.value)
    
    def div(this, other: 'Galois') -> 'Galois':
        return Galois(this.P, this.value * this._inverse(other.value))
    
    def pow(this, other: 'Galois') -> 'Galois':
        return Galois(this.P, pow(this.value, other.value, this.P))
    
    